sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosPermissionProperties.java

```diff
@@ -93,63 +189,138 @@ public CosmosPermissionProperties setPermissionMode(PermissionMode permissionMod
      * @return the partition key.
      */
     public PartitionKey getResourcePartitionKey() {
-        return this.permission.getResourcePartitionKey();
+        return this.resourcePartitionKey;
     }
 
     /**
      * Sets the resource partition key associated with this permission object.
      *
      * @param partitionKey the partition key.
-     * @return the current {@link CosmosPermissionProperties} object
+     * @return the current {@link CosmosPermissionProperties} object.
      */
     public CosmosPermissionProperties setResourcePartitionKey(PartitionKey partitionKey) {
-        this.permission.setResourcePartitionKey(partitionKey);
+        this.resourcePartitionKey = partitionKey;
+
+        // Following is required by permission replace (PUT) scenario.
+        if (this.permission != null) {
+            this.permission.setResourcePartitionKey(partitionKey);
+            this.resourceToken = null;
+        }
+
         return this;
     }
 
     Resource getResource() {
         return this.permission;
     }
 
+    String getResourcePath(String databaseName) {
```

Comment:
can the logic in Utils be used to avoid redoing similar logic here?

Reply:
I could not find any... there are Utils APIs that do some of the splitting for a resource path, not applicable for this case though.

Reply:
How about
`PathHelper#generatePathForNameBased()` 
https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PathsHelper.java#L30

There are other useful methods there too. can they be used instead of duplicating the logic?

Reply:
generatePathForNameBased(0 an be used to a certain degree but it requires multiple calls and passing it the Resource argument which is not currently available. IMHO the code will look a lot messier and hard to understand.

For the constant/hardcoded names we already used the predefined final strings; we do similar things in Utils for certain type of resources. I think the current implementation has the right balance and it's easy to check what the expected results should look like in case of debugging.